base/compiler/inferencestate.jl

```diff
@@ -184,7 +184,7 @@ mutable struct InferenceState
         #       are stronger than the inbounds assumptions, since the latter
         #       requires dynamic reachability, while the former is global).
         inbounds = inbounds_option()
-        noinbounds = inbounds === :on || (inbounds === :default && !any_inbounds(code))
+        noinbounds = inbounds === :on || (inbounds === :default && all(flag::UInt8->iszero(flag&IR_FLAG_INBOUNDS), src.ssaflags))
```

Comment:
On further thought, we probably need to taint :consistent-cy _always_ when we see `:inbounds` or `:boundscheck` expressions since e.g. a `-check-bounds=no` session may use cache that is generated in `-check-bounds=yes`?
In other word, I think we can't do this until we separate cache with boundscheck flag.

Reply:
But they are separated now I think. Since the Pkg image stuff.

Reply:
With `--pkgimage=yes` the caches are split, with `--pkgimage=no` they aren't.

Reply:
Okay, then I think we can safely skip tainting `:consistent`-cy when abstract interpret sees `:inbounds`/`:boundscheck` exprs if `--pkgimages=yes && --checkbounds=yes`.

A next question would be: is it fast enough to check `JLOptions().[use_pkgimages|check_bounds]` during abstract interpret?

Reply:
They are split with pkgimages, but the sysimage itself is still not split, so we can't do this.

Reply:
So we need to taint `:consistent`-cy always upon :inbounds or :boundscheck expressions, and probably want to remove these `:noinbounds` effect bits until we completely separate cache by boundscheck flag?